### ISSUE TITLE
Use jsdoc convetions to enable autocompletion for editors

### ISSUE THREAD

**Issue by noandrea** [aeternity/aepp-sdk-js#736]
**I'm submitting a ...**
  - [ ] bug report
  - [x] feature request
  - [ ] support request => Please do not submit support request here, see note at the top of this template.

Use [type](https://jsdoc.app/tags-type.html) and [typedef](https://jsdoc.app/tags-typedef.html) from jsdoc 

* **What is the motivation / use case for changing the behavior?**

* **Please tell us about your environment:**
  
  - Node Version: v0.0.0
  - Protocol Version: 1
  - Compiler version: v0.0.0
  - VM Version: aevm | fate
  - SDK Version: v0.0.0
  - Environment: browser | nodejs

* **Other information** (e.g. detailed explanation, stacktraces, related issues, suggestions how to fix, links for us to have context, eg. forum, telegram, etc)

**Comment by noandrea** [comment]
also add typescript type file ( http://definitelytyped.org/ )

**Comment by arvinxx** [comment]
is there a `.d.ts` file？

**Comment by the-icarus** [comment]
Is there still work done on this issue? Having autocomplete support on IDEs would be really helpful for all the developers working with the aepp-sdk.
I did a quick evaluation and since there are already a lot of jsdoc comments attached to the functions. It is quite easy to create the `d.ts` files.
I used the following tsconfig to create `d.ts` files located next to all the `.js` files:
```json
{
    "compilerOptions": {
        "emitDeclarationOnly": true,
        "declaration": true,
        "allowJs": true,
        "outDir": "typings"
    },
    "files": [
        "es/index.js"
    ],
    "include": [
        "es/**/*.js"
    ],
    "exclude": [
        "node_modules/**/*"
    ]
}
```
This will emit `d.ts` files to the typings folder

**Comment by the-icarus** [comment]
Small addition: Running the config without `outDir` will place the `d.ts` files next to the `js` files. VS Code will then automatically detect the definition file when importing the module. This will provide IntelliSense for the aepp-sdk

**Comment by nduchak** [comment]
Hey @the-icarus ,
Thanks for your investigation, but the problem is that that typing is generated properly,
because of using `ramda`, but it can be a good starting point for manual adjustments

**Comment by the-icarus** [comment]
Hey @nduchak,

That’s true. For some functions like `toAettos` in the `amount-formatter.js` the generates `.d.ts` is quite good and has the correct types. For others we often get `any` as type which is not useful at all. I will investigate further and check if this can be adjusted using JSDoc comments.

**Comment by nduchak** [comment]
I made some investigation of that and the main problem was how to properly write jsdocs for Stampit library (we ate using  this instead of classic class inheritance)

**Comment by the-icarus** [comment]
Yes this was also my thought. I will try to find a way to improve JSDoc comments for better type generation.

**Comment by the-icarus** [comment]
Hi @nduchak ,
I made some further investigations.
Basically, you are right. Using stampit makes it quite hard to generate types.
However, we can use JSDoc comments and reference/define types in there. This will be some effort since we have to define own types for the stamps.

I made a quick sample for the spendTx function. The original file is located at `es/tx/index.js`
First I added the type definition for the options:
```
/**
 * @typedef SpendTxOptions
 * @property {string?} sender
 * @property {string} recipientId
 * @property {number} amount
 * @property {number?} fee
 * @property {number?} ttl
 * @property {number?} nonce
 * @property {string?} payload
 */
```
Then I added the options to the function definition:
```
/**
 * Create a `spend_tx` transaction
 * @function spendTx
 * @instance
 * @abstract
 * @category async
 * @rtype ({sender?: String, recipientId: String, amount: Number, fee?: Number, ttl?: Number, nonce?: Number, payload?: String}) => tx: Promise[String]
 * @callback spendTx
 * @param {SpendTxOptions} options - The object to extract properties from
 * @return {Promise<string>} `spend_tx` transaction
 */
```
To make the function type available in `d.ts` we need to add `@callback` annotation.
We can now use this definition when for all stamps which implement `spendTx`-Function.

As a starting point we could add the typescript compiler for `d.ts` generation and add more types from time to time.
I think having `d.ts` files and IDE type support will improve developer experience and simplify adoption.

What do you think?

**Comment by nduchak** [comment]
First of all thanks for you help🙏 And I am totally agree with you.
The one question do you find out how to comment the stamps?

**Comment by the-icarus** [comment]
Yes and no.
You can just directly annotate with type
```
/**
* @type {{
*   spendTx: spendTx
* }}
*/
```
This indicates that the stamp is of a type containing a property spendTx of type spendTx (function from previous post).
But I still try to figure out if we can merge types (similar to what you are doing with stampit) to compose new types with ease.

**Comment by the-icarus** [comment]
@nduchak So I found out, that you can compose types with simple `&` operator.
Think of two types Oracle and AENS which define their functions. When we now compose the Tx type which is a mix of Oralce and AENS we can define the type like this:
```
/**
 * Basic Tx Stamp
 *
 * Attempting to create instances from the Stamp without overwriting all
 * abstract methods using composition will result in an exception.
 *
 * Tx is one of the three basic building blocks of an
 * {@link module:@aeternity/aepp-sdk/es/ae--Ae} client and provides methods to
 * create aeternity transactions.
 * @function
 * @alias module:@aeternity/aepp-sdk/es/tx
 * @rtype Stamp
 * @param {Object} [options={}] - Initializer object
 * @return {Object} Tx instance
 * @type Oracle & ANES
 * @example Tx()
 */
```
The relevant part is the `@type` annotation. We can also add additional custom fields like this:
```
@type Oracle & AENS & {spendTx: spendTx}
```
The result object will get IntelliSense for all Oracle functions/properties, AENS functions/properties and spendTx function.

**Comment by nduchak** [comment]
Nice. Will test and after start working on PR with basic typings and then will adjust the JsDocs one by one

**Comment by the-icarus** [comment]
@nduchak I did some additional testing and had some findings I would like to share with you.
I tried to propagate a type on Transaction stamp and see IntelliSense when using RpcAepp.
I added following files to `es/tx/tx.js`
```Typescript
/**
* [...]
 * @typedef Transaction
 * @type {{
 *  spendTx: spendTx
 * }}
 * [...]
 */
```
In `es/ae/aepp.js` I added a type: 
```Typescript
/***
 * @callback RpcAepp
 * @param {object} options
 * @return {Promise<import('../tx/tx').Transaction>}
 */
```
And added the type to RpcAepp stamp declaration:
```Typescript
/**
 * RpcAepp Stamp
 * 
 * @type RpcAepp
 */
export const RpcAepp = Ae.compose(Tx, Oracle, Contract, Aens, Chain, AeppRpc)
```
Afterwards, I got IntelliSense on RpcAepp.

*In case you reference types from other files you can use the `import` syntax shown in the second code block*

**Comment by the-icarus** [comment]
@nduchak are there any updates on this issue? I see #867 is also related to docs. I still think providing autocomplete and type support will improve developer experience. In case there are Ressource/budget limitations we can try to prioritize and I may have time to help with implementation.

**Comment by davidyuk** [comment]
@the-icarus Unfortunately, Nazar is not working on the SDK anymore. This is maintained by Superhero developers (primary me) on demand. From my perspective, providing some type definitions in a short term is fine (feel free to provide a PR with corresponding functionality, even if it is not complete), but in the long run, I would like would drop Stamps, switch to traditional OOP, and start writing on TS 🚀

**Comment by the-icarus** [comment]
@davidyuk Thanks for the update. I agree on dropping the stamps and switch to native TypeScript. I think it does not make much sense to include some typings when the plan is to switch to TypeScript. Have you already created a strategy how we can achieve the switch to TypeScript? I might can support you with the transition or in creating a transition strategy.

You also mention there is only on-demand maintenance. Did I get that right that currently there is no dedicated maintainer for the SDK?

**Comment by the-icarus** [comment]
This will be solved by switching to typescript. See #1113
